OWS's Flask app creation isn't  thread safe

After having OWS 1.9.6 running perfectly in the test instance, I dumped my database and deployed the same images + database dump in the production instance, and my OWS errors re-appeared. After various attempts to debug, I finally realized that multiple threads are calling alembic to see if the database revision is the latest, and when they run their `__exit__()` methods they try to delete keys from the same dictionary so they fail with "script", "config", or other strange messages.

Here is a backtrace that I got after adding more logging calls:
```
[2025-12-09 13:08:42,359] [ERROR] 'config'
Traceback (most recent call last):
  File "/app/lib/python3.12/site-packages/datacube_ows/ogc.py", line 182, in ogc_svc_impl
    return version_support.router(nocase_args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/lib/python3.12/site-packages/datacube_ows/utils.py", line 38, in log_wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/app/lib/python3.12/site-packages/datacube_ows/wms.py", line 29, in handle_wms
    return get_map(nocase_args)
           ^^^^^^^^^^^^^^^^^^^^
  File "/app/lib/python3.12/site-packages/datacube_ows/utils.py", line 38, in log_wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/app/lib/python3.12/site-packages/datacube_ows/data.py", line 107, in get_map
    n_datasets = stacker.n_datasets()
                 ^^^^^^^^^^^^^^^^^^^^
  File "/app/lib/python3.12/site-packages/datacube_ows/loading.py", line 200, in n_datasets
    return self._layer.ows_index().count(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/lib/python3.12/site-packages/datacube_ows/index/api.py", line 236, in inner
    dc = Datacube(env=dcl.local_env)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/lib/python3.12/site-packages/datacube/api/core.py", line 139, in __init__
    self.index = index_connect(
                 ^^^^^^^^^^^^^^
  File "/app/lib/python3.12/site-packages/datacube/index/_api.py", line 43, in index_connect
    return index_driver.connect_to_index(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/lib/python3.12/site-packages/datacube/index/abstract/_index.py", line 410, in connect_to_index
    return cls.index_class().from_config(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/lib/python3.12/site-packages/datacube/index/postgis/index.py", line 136, in from_config
    db = PostGisDb.from_config(
         ^^^^^^^^^^^^^^^^^^^^^^
  File "/app/lib/python3.12/site-packages/datacube/drivers/postgis/_connections.py", line 82, in from_config
    return PostGisDb.create(
           ^^^^^^^^^^^^^^^^^
  File "/app/lib/python3.12/site-packages/datacube/drivers/postgis/_connections.py", line 109, in create
    if not _core.schema_is_latest(engine):
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/lib/python3.12/site-packages/datacube/drivers/postgis/_core.py", line 249, in schema_is_latest
    with EnvironmentContext(cfg, scriptdir) as env_ctx:
  File "/app/lib/python3.12/site-packages/alembic/runtime/environment.py", line 217, in __exit__
    self._remove_proxy()
  File "/app/lib/python3.12/site-packages/alembic/util/langhelpers.py", line 87, in _remove_proxy
    del globals_[attr_name]
        ~~~~~~~~^^^^^^^^^^^
KeyError: 'config'
```
I think the solution is to do what Explorer already does, open the datacube index without validation (https://github.com/opendatacube/datacube-explorer/blob/18b29e3d381c6b3881567f6262a1fd4ede808941/cubedash/_model.py#L63) and have a `create_app()` function to create the Flask app that is called from the Dockerfile (https://github.com/opendatacube/datacube-explorer/blob/18b29e3d381c6b3881567f6262a1fd4ede808941/Dockerfile#L122).

--- a/Dockerfile
+++ b/Dockerfile
@@ -108,4 +108,16 @@ ENV GDAL_DISABLE_READDIR_ON_OPEN="EMPTY_DIR" \
     PATH=/app/bin:$PATH
 
 ENTRYPOINT ["/usr/local/bin/remap-user.sh"]
-CMD ["gunicorn", "-b", "0.0.0.0:8000", "--workers=3", "-k", "gevent", "--timeout", "121", "--pid", "/home/ubuntu/gunicorn.pid", "--log-level", "info", "--worker-tmp-dir", "/dev/shm", "--config", "python:datacube_ows.gunicorn_config", "datacube_ows.wsgi"]
+CMD [\
+    "gunicorn", \
+    "-b", "0.0.0.0:8000", \
+    "--workers=3", \
+    "-k", \
+    "gevent", \
+    "--timeout", "121", \
+    "--pid", "/home/ubuntu/gunicorn.pid", \
+    "--log-level", "info", \
+    "--worker-tmp-dir", "/dev/shm", \
+    "--config", "python:datacube_ows.gunicorn_config", \
+    "datacube_ows.startup_utils:create_app()"\
+]

--- a/datacube_ows/__init__.py
+++ b/datacube_ows/__init__.py
@@ -10,3 +10,7 @@ except ImportError:
     # Default version number.
     # Will only be used when running datacube-ows direct from source code (not properly installed)
     __version__ = "1.9.6"
+
+from .startup_utils import create_app
+
+__all__ = ["__version__", "create_app"]

--- a/datacube_ows/ogc.py
+++ b/datacube_ows/ogc.py
@@ -3,13 +3,12 @@
 #
 # Copyright (c) 2017-2024 OWS Contributors
 # SPDX-License-Identifier: Apache-2.0
-
+import logging
 import sys
 import traceback
 from logging import Logger
-from time import monotonic
 
-from flask import g, render_template, request
+from flask import render_template, request
 
 from datacube_ows import __version__
 from datacube_ows.http_utils import (
@@ -21,75 +20,19 @@ from datacube_ows.http_utils import (
 from datacube_ows.index.api import ows_index
 from datacube_ows.legend_generator import create_legend_for_style
 from datacube_ows.ogc_exceptions import OGCException, WMSException
-from datacube_ows.ows_configuration import OWSConfig, get_config
+from datacube_ows.ows_configuration import get_config
 from datacube_ows.protocol_versions import SupportedSvc, supported_versions
-from datacube_ows.startup_utils import (
-    initialise_aws_credentials,
-    initialise_babel,
-    initialise_debugging,
-    initialise_flask,
-    initialise_ignorable_warnings,
-    initialise_logger,
-    initialise_prometheus,
-    initialise_sentry,
-    parse_config_file,
-    proxy_fix,
-)
 from datacube_ows.wcs1 import WCS_REQUESTS
 from datacube_ows.wms import WMS_REQUESTS
 
-# Logging intialisation
-_LOG: Logger = initialise_logger()
-initialise_ignorable_warnings()
-
-# Initialisation of external libraries - controlled by environment variables.
-initialise_debugging(_LOG)
-initialise_sentry(_LOG)
-initialise_aws_credentials(_LOG)
-
-# Prepare parsed configuration object
-cfg: OWSConfig | None = parse_config_file()
-
-# Initialise Flask
-app = initialise_flask(__name__)
-
-babel = initialise_babel(cfg, app)
-
-# Initialisation of external libraries that depend on Flask
-# (controlled by environment variables)
-metrics = initialise_prometheus(app, _LOG)
-
-# Add middleware to fix proxy headers, controlled by environment variables
-app = proxy_fix(app, _LOG)
+_LOG: Logger = logging.getLogger(__name__)
 
 # Protocol/Version lookup table
 OWS_SUPPORTED: dict[str, SupportedSvc] = supported_versions()
 
-# Prometheus Metrics
-prometheus_ows_ogc_metric = metrics.histogram(
-    "ows_ogc",
-    "Summary by OGC request protocol, version, operation, layer, and HTTP Status",
-    labels={
-        "query_request": lambda: request.args.get("request", "NONE").upper(),
-        "query_service": lambda: request.args.get("service", "NONE").upper(),
-        "query_version": lambda: request.args.get("version"),
-        "query_layer": lambda: (
-            request.args.get("query_layers")  # WMS GetFeatureInfo
-            or request.args.get("layers")  # WMS
-            or request.args.get("layer")  # WMTS
-            or request.args.get("coverage")  # WCS 1.x
-            or request.args.get("coverageid")  # WCS 2.x
-        ),
-        "status": lambda r: r.status_code,
-    },
-)
-
 
 # Flask Routes
-
-
-@app.route("/")
-@prometheus_ows_ogc_metric
+# (Note that actual route declarations take place in startup_utils.py)
 def ogc_impl():
     # pylint: disable=too-many-branches
     nocase_args = lower_get_args()
@@ -192,28 +135,18 @@ def ogc_svc_impl(svc):
         return ogc_e.exception_response(traceback=traceback.extract_tb(tb))
 
 
-@app.route("/wms")
-@prometheus_ows_ogc_metric
 def ogc_wms_impl():
     return ogc_svc_impl("wms")
 
 
-@app.route("/wmts")
-@prometheus_ows_ogc_metric
 def ogc_wmts_impl():
     return ogc_svc_impl("wmts")
 
 
-@app.route("/wcs")
-@prometheus_ows_ogc_metric
 def ogc_wcs_impl():
     return ogc_svc_impl("wcs")
 
 
-@app.route("/ping")
-@metrics.summary(
-    "ows_heartbeat_pings", "Ping durations", labels={"status": lambda r: r.status}
-)
 def ping() -> tuple[str, int, dict[str, str]]:
     dbs_ok = {
         name: ows_index(dc).check_db_access(dc)
@@ -239,16 +172,6 @@ def ping() -> tuple[str, int, dict[str, str]]:
     )
 
 
-@app.route("/legend/<string:layer>/<string:style>/legend.png")
-@metrics.histogram(
-    "ows_legends",
-    "Legend query durations",
-    labels={
-        "layer": lambda: request.path.split("/")[2],
-        "style": lambda: request.path.split("/")[3],
-        "status": lambda r: r.status,
-    },
-)
 def legend(layer, style, dates=None):
     # pylint: disable=redefined-outer-name
     cfg = get_config()
@@ -268,36 +191,3 @@ def legend(layer, style, dates=None):
     if not img:
         return "Unknown Style", 404, resp_headers({"Content-Type": "text/plain"})
     return img
-
-
-# Flask middleware
-
-
-@app.before_request
-def start_timer() -> None:
-    # pylint: disable=assigning-non-slot
-    g.ogc_start_time = monotonic()
-
-
-@app.after_request
-def log_time_and_request_response(response):
-    time_taken = int((monotonic() - g.ogc_start_time) * 1000)
-    # request.environ.get('HTTP_X_REAL_IP') captures requester ip on a local docker container via gunicorn
-    if request.environ.get("HTTP_X_REAL_IP"):
-        ip = request.environ.get("HTTP_X_REAL_IP")
-    # request.environ.get('HTTP_X_FORWARDED_FOR') captures request IP forwarded by ingress/loadbalancer
-    elif request.environ.get("HTTP_X_FORWARDED_FOR"):
-        ip = request.environ.get("HTTP_X_FORWARDED_FOR")
-    # request.environ.get('REMOTE_ADDR') is standard internal IP address
-    elif request.environ.get("REMOTE_ADDR"):
-        ip = request.environ.get("REMOTE_ADDR")
-    else:
-        ip = "Not found"
-    _LOG.info(
-        "ip: %s request: %s returned status: %d and took: %d ms",
-        ip,
-        request.url,
-        response.status_code,
-        time_taken,
-    )
-    return response

--- a/datacube_ows/startup_utils/__init__.py
+++ b/datacube_ows/startup_utils/__init__.py
@@ -9,11 +9,12 @@ import os
 import warnings
 from collections.abc import Callable
 from logging import Logger
+from time import monotonic
 from typing import Any, TypeVar
 
 from botocore.credentials import RefreshableCredentials
 from datacube.utils.aws import configure_s3_access
-from flask import Flask, request
+from flask import Flask, g, request
 from rasterio.errors import NotGeoreferencedWarning
 
 from datacube_ows.ows_configuration import OWSConfig, get_config
@@ -29,20 +30,21 @@ __all__ = [
     "initialise_flask",
     "initialise_prometheus",
     "CredentialManager",
+    "create_app",
 ]
 
 
 def initialise_logger(name: str | None = None) -> Logger:
     handler = logging.StreamHandler()
     handler.setFormatter(logging.Formatter("[%(asctime)s] [%(levelname)s] %(message)s"))
-    _LOG = logging.getLogger(name)
-    _LOG.addHandler(handler)
+    log = logging.getLogger(name)
+    log.addHandler(handler)
     # If invoked using Gunicorn, link our root logger to the gunicorn logger
     # this will mean the root logs will be captured and managed by the gunicorn logger
     # allowing you to set the gunicorn log directories and levels for logs
     # produced by this application
-    _LOG.setLevel(logging.getLogger("gunicorn.error").getEffectiveLevel())
-    return _LOG
+    log.setLevel(logging.getLogger("gunicorn.error").getEffectiveLevel())
+    return log
 
 
 def initialise_ignorable_warnings() -> None:
@@ -210,7 +212,7 @@ def parse_config_file(log: Logger | None = None) -> OWSConfig | None:
 
 
 def initialise_flask(name: str) -> Flask:
-    app_path = os.path.dirname(os.path.abspath(__file__))
+    app_path = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
     return Flask(
         name.split(".")[0], template_folder=os.path.join(app_path, "templates")
     )
@@ -288,3 +290,114 @@ def initialise_babel(cfg, app: Flask) -> object | None:
             configure_jinja=False,
         )
     return None
+
+
+def create_app() -> Flask:
+    log = initialise_logger()
+    cfg = parse_config_file(log)
+    initialise_ignorable_warnings()
+    initialise_debugging(log)
+    initialise_sentry(log)
+    initialise_aws_credentials(log)
+    app = initialise_flask(__name__)
+    initialise_babel(cfg, app)
+    metrics = initialise_prometheus(app, log)
+    ows_ogc_metric = metrics.histogram(
+        "ows_ogc",
+        "Summary by OGC request protocol, version, operation, layer, and HTTP Status",
+        labels={
+            "query_request": lambda: request.args.get("request", "NONE").upper(),
+            "query_service": lambda: request.args.get("service", "NONE").upper(),
+            "query_version": lambda: request.args.get("version"),
+            "query_layer": lambda: (
+                request.args.get("query_layers")  # WMS GetFeatureInfo
+                or request.args.get("layers")  # WMS
+                or request.args.get("layer")  # WMTS
+                or request.args.get("coverage")  # WCS 1.x
+                or request.args.get("coverageid")  # WCS 2.x
+            ),
+            "status": lambda r: r.status_code,
+        },
+    )
+    proxy_fix(app, log)
+
+    # Declare routes
+    from datacube_ows.ogc import (
+        legend,
+        ogc_impl,
+        ogc_wcs_impl,
+        ogc_wms_impl,
+        ogc_wmts_impl,
+        ping,
+    )
+
+    @app.route("/")
+    @ows_ogc_metric
+    def main_route():
+        return ogc_impl()
+
+    @app.route("/wms")
+    @ows_ogc_metric
+    def wms_route():
+        return ogc_wms_impl()
+
+    @app.route("/wmts")
+    @ows_ogc_metric
+    def wmts_route():
+        return ogc_wmts_impl()
+
+    @app.route("/wcs")
+    @ows_ogc_metric
+    def wcs_route():
+        return ogc_wcs_impl()
+
+    @app.route("/ping")
+    @metrics.summary(
+        "ows_heartbeat_pings", "Ping durations", labels={"status": lambda r: r.status}
+    )
+    def ping_route():
+        return ping()
+
+    @app.route("/legend/<string:layer>/<string:style>/legend.png")
+    @metrics.histogram(
+        "ows_legends",
+        "Legend query durations",
+        labels={
+            "layer": lambda: request.path.split("/")[2],
+            "style": lambda: request.path.split("/")[3],
+            "status": lambda r: r.status,
+        },
+    )
+    def legend_route(layer, style):
+        return legend(layer, style)
+
+    # Configure Flask Middleware
+    @app.before_request
+    def start_timer() -> None:
+        # pylint: disable=assigning-non-slot
+        g.ogc_start_time = monotonic()
+
+    @app.after_request
+    def log_time_and_request_response(response):
+        time_taken = int((monotonic() - g.ogc_start_time) * 1000)
+        # request.environ.get('HTTP_X_REAL_IP') captures requester ip on a local docker container via gunicorn
+        if request.environ.get("HTTP_X_REAL_IP"):
+            ip = request.environ.get("HTTP_X_REAL_IP")
+        # request.environ.get('HTTP_X_FORWARDED_FOR') captures request IP forwarded by ingress/loadbalancer
+        elif request.environ.get("HTTP_X_FORWARDED_FOR"):
+            ip = request.environ.get("HTTP_X_FORWARDED_FOR")
+        # request.environ.get('REMOTE_ADDR') is standard internal IP address
+        elif request.environ.get("REMOTE_ADDR"):
+            ip = request.environ.get("REMOTE_ADDR")
+        else:
+            ip = "Not found"
+        log.info(
+            "ip: %s request: %s returned status: %d and took: %d ms",
+            ip,
+            request.url,
+            response.status_code,
+            time_taken,
+        )
+        return response
+
+    return app

--- a/datacube_ows/wsgi.py
+++ b/datacube_ows/wsgi.py
@@ -10,24 +10,22 @@ import os
 import sys
 
 # This is the directory of the source code that the web app will run from
-sys.path.append("/opt")
+sys.path.append("/src")
 
 # The location of the datacube config file.
-if os.path.isfile("/opt/odc/.datacube.conf.local"):
-    os.environ.setdefault("ODC_CONFIG_PATH", "/opt/odc/.datacube.conf.local")
+if os.path.isfile("/src/odc/.datacube.conf.local"):
+    os.environ.setdefault("ODC_CONFIG_PATH", "/src/odc/.datacube.conf.local")
 
-from datacube_ows import __version__
+from datacube_ows import __version__, create_app
 
-from datacube_ows.ogc import app  # isort:skip
-
-application = app
+application = create_app()
 
 
 def main() -> None:
     if "--version" in sys.argv:
         print("Open Data Cube Open Web Services (datacube-ows) version", __version__)
         exit(0)
-    app.run()
+    application.run()
 
 
 if __name__ == "__main__":

--- a/docker-compose.yaml
+++ b/docker-compose.yaml
@@ -31,7 +31,7 @@ services:
       POSTGRES_HOSTNAME: ${POSTGRES_HOSTNAME}
       SERVER_DB_USERNAME: ${SERVER_DB_USERNAME}
       # Path from the PYTHONPATH to the config object (default PYTHONPATH is /src)
-      PYTHONPATH: ${PYTHONPATH}
+      PYTHONPATH: /src:${PYTHONPATH}
       DATACUBE_OWS_CFG: ${DATACUBE_OWS_CFG}
       AWS_DEFAULT_REGION: ${AWS_REGION}
       # Talk to AWS without using credentials
@@ -39,7 +39,7 @@ services:
       # Enable Metrics
       PROMETHEUS_MULTIPROC_DIR: ${PROMETHEUS_MULTIPROC_DIR}
       # Dev flags
-      FLASK_APP: /src/datacube_ows/ogc.py
+      FLASK_APP: "datacube_ows.startup_utils:create_app"
       FLASK_ENV: ${FLASK_ENV}
       PYDEV_DEBUG: "${PYDEV_DEBUG}"
       SENTRY_DSN: "${SENTRY_DSN}"
@@ -47,5 +47,5 @@ services:
       - ${OWS_CFG_DIR}:${OWS_CFG_MOUNT_DIR}
       - ./:/src/
       - ./artifacts:/mnt/artifacts
-    restart: always
+    # restart: always
     command: ["flask", "run", "--host=0.0.0.0", "--port=8000"]

--- a/integration_tests/conftest.py
+++ b/integration_tests/conftest.py
@@ -12,15 +12,21 @@ pytest_plugins = ["helpers_namespace"]
 import pytest
 from click.testing import CliRunner
 from datacube.cfg import ODCConfig
+from flask import Flask
 from pytest_localserver.http import WSGIServer
 
-from datacube_ows import ogc
-from datacube_ows.ogc import app
+from datacube_ows import create_app
+
+
+@pytest.fixture(scope="session")
+def flask_app() -> Generator[Flask]:
+    app = create_app()
+    yield app
 
 
 @pytest.fixture
-def flask_client() -> Generator[WSGIServer]:
-    with app.test_client() as client:
+def flask_client(flask_app) -> Generator[WSGIServer]:
+    with flask_app.test_client() as client:
         yield client
 
 
@@ -29,7 +35,7 @@ class generic_obj:
 
 
 @pytest.fixture(scope="session")
-def ows_server(request) -> WSGIServer:
+def ows_server(request, flask_app) -> WSGIServer:
     """
     Run the OWS server for the duration of these tests
     """
@@ -38,7 +44,7 @@ def ows_server(request) -> WSGIServer:
         server = generic_obj()
         server.url = external_url
     else:
-        server = WSGIServer(port="5000", application=ogc.app)
+        server = WSGIServer(port="5000", application=flask_app)
         server.start()
         request.addfinalizer(server.stop)
 

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -22,9 +22,9 @@ from tests.utils import coords, dim1_da, dim1_da_time, dummy_da
 @pytest.fixture
 def flask_client(monkeypatch) -> Generator[flask.Flask]:
     monkeypatch.setenv("DEFER_CFG_PARSE", "yes")
-    from datacube_ows.ogc import app
+    from datacube_ows.startup_utils import create_app
 
-    with app.test_client() as client:
+    with create_app().test_client() as client:
         yield client
 
 

